hcl/fmt.go

```diff
@@ -274,99 +263,114 @@ func fmtListExpr(tokens hclwrite.Tokens) (hclwrite.Tokens, int) {
 		if isHeredoc(element) {
 			newTokens = append(newTokens, newlineToken())
 		}
-		newTokens = append(newTokens, commaToken(), newlineToken())
+
+		// On some scenarios like { [] = etc, ... } we need to avoid double commas
+		if newTokens[len(newTokens)-1].Type != hclsyntax.TokenComma {
+			newTokens = append(newTokens, commaToken())
+		}
+		newTokens = append(newTokens, newlineToken())
 	}
 
 	newTokens = append(newTokens, closeBracketToken())
-	elemNextPos++
+	elemIndex++
 
 	// Handling ["one"][0] and things like [[0]%[0]]
 	// We can also have newlines when dealing with operations
 
 	logger.Trace().Msg("checking if formatted list has operators/index access")
 
-	skipNls()
-
-	if elemNextPos == len(tokens) {
+	if elemIndex == len(tokens) {
 		logger.Trace().Msg("no more tokens, returning formatted list")
-		return newTokens, elemNextPos
+		return newTokens, elemIndex
 	}
 
 	// We are handling things like this:
 	// var = [[] # c1\n\n #c\n [*]]
 	// We need to keep any comments after the immediate end of the list
 	// We don't keep the extra newlines, only newlines belonging to comments themselves.
-	hasCommentOrNl := true
-	for hasCommentOrNl {
-		switch tokens[elemNextPos].Type {
+	searchCommentOrNl := true
+	hasNewlineBeforeToken := false
+
+	for searchCommentOrNl {
+		switch tokens[elemIndex].Type {
 		case hclsyntax.TokenComment:
 			logger.Trace().Msg("found comment after end of list, adding token")
-			newTokens = append(newTokens, tokens[elemNextPos])
-			elemNextPos++
+			newTokens = append(newTokens, tokens[elemIndex])
+			elemIndex++
 		case hclsyntax.TokenNewline:
 			logger.Trace().Msg("found newline after end of list, ignoring")
-			elemNextPos++
+			hasNewlineBeforeToken = true
+			elemIndex++
 		default:
-			hasCommentOrNl = false
+			searchCommentOrNl = false
 		}
 	}
```

Comment:
hmm I don't think it reverses, it ignores all newlines keeping only the comments. For  `NL COMMENT` It will first check the NL and skip it (increments index), and then when it selects the comment it will add the comment to the newTokens, so newTokens will only have `COMMENT` inside.

So it would work for something like:

```
{
  a = value
  // some comment
  b = value
}
```

We filter away the newline + keep the comment and hasNewlineBeforeToken is used to identify that in this case the next token `b` is actually the next key = value pair and will need a newline after the end of the list.

Of course I could also be wrong on this :laughing: . In the end this is related to handling the object without commas without causing this:

```
{
  a = [
  ] b = 666
}
```

There could be alternative approaches to achieve that.

Reply:
yes, because the `hasNewlineBeforeToken` is checked later in the line 334 but the comment there says it handles `[]\n[0]` but it's also true for `[]\n/* comment */[0]` which makes `[]\n/* comment */\n[0]` but it looks ok to me..
feel free to ignore this.

Reply:
I will do some digging on this again, I could have forgot something, thanks for pointing it out :100:

Reply:
Double checked that it will remove the newlines and keep the comments.